sklearn/utils/openmp_helpers.pyx

```diff
@@ -0,0 +1,26 @@
+cimport openmp
+from joblib import effective_n_jobs
+
+
+cpdef _openmp_effective_n_threads(n_threads=None):
+    """Determine the effective number of threads used for parallel OpenMP calls
+
+    - For ``n_threads = None``, returns the minimum between
+      openmp.omp_get_max_threads() and joblib.effective_n_jobs(-1).
+    - For ``n_threads > 0``, use this as the maximal number of threads for
+      parallel OpenMP calls.
+    - For ``n_threads < 0``, use the maximal number of threads minus
+      ``|n_threads + 1|``.
+    - Raise a ValueError for ``n_threads = 0``.
+    """
+    if n_threads == 0:
+        raise ValueError("n_threads = 0 is invalid")
+
+    max_threads = min(openmp.omp_get_max_threads(), effective_n_jobs(-1))
```

Comment:
Is there a case when `openmp.omp_get_max_threads()` would be lower than `joblib.effective_n_jobs(-1)` ?

Reply:
yes, if a user has first set `OMP_NUM_THREADS=1` for instance

Reply:
Also, some implementation of OpenMP return the number of __physical cores__ instead of virtual cores, which is typically lower than what is reported by `joblib.effective_n_jobs`.

Reply:
Interesting. This does not always seem to be the case though. Especially not libgomp. So it would be great if loky could have an option to report the number of physical cores as commented here :https://github.com/scikit-learn/scikit-learn/pull/14196#pullrequestreview-290041629